File: packet_seaquence/log/go_src/models/protocol_analyzer/ipv4.go
// Package protocol_analyzer provides IPv4 protocol analyzer
package protocol_analyzer

import (
	"fmt"
)

// AnalyzeIPv4 implements the ProtocolAnalyzer interface for IPv4
type AnalyzeIPv4 struct {
	BaseProtocolAnalyzer
}

// NewAnalyzeIPv4 creates a new IPv4 analyzer
func NewAnalyzeIPv4(packet Packet) *AnalyzeIPv4 {
	return &AnalyzeIPv4{
		BaseProtocolAnalyzer: NewBaseProtocolAnalyzer(packet),
	}
}

// Analyze extracts IPv4 information from a packet
func (a *AnalyzeIPv4) Analyze() map[string]interface{} {
	if !a.Pkt.LayerExists("ip") {
		return nil
	}

	ipv4Info := make(map[string]interface{})
	
	if srcIP, ok := a.Pkt.GetField("ip", "src"); ok {
		ipv4Info["src_ip"] = srcIP
	}
	
	if dstIP, ok := a.Pkt.GetField("ip", "dst"); ok {
		ipv4Info["dst_ip"] = dstIP
	}
	
	if ttl, ok := a.Pkt.GetField("ip", "ttl"); ok {
		ipv4Info["ttl"] = ttl
	}
	
	if flags, ok := a.Pkt.GetField("ip", "flags"); ok {
		ipv4Info["flags"] = flags
	}
	
	if length, ok := a.Pkt.GetField("ip", "len"); ok {
		ipv4Info["length"] = length
	}
	
	return ipv4Info
}

// GetDisplayInfo returns formatted IPv4 information
func (a *AnalyzeIPv4) GetDisplayInfo() string {
	info := a.Analyze()
	if info == nil {
		return "IPv4情報なし"
	}
	
	srcIP, _ := info["src_ip"].(string)
	dstIP, _ := info["dst_ip"].(string)
	ttl, _ := info["ttl"].(string)
	
	return fmt.Sprintf("IPv4 %s -> %s (TTL=%s)", 
		getOrDefault(srcIP, "?"),
		getOrDefault(dstIP, "?"),
		getOrDefault(ttl, "?"))
}

// GetSummary returns a summary of IPv4 information
func (a *AnalyzeIPv4) GetSummary() string {
	info := a.Analyze()
	if info == nil {
		return "IPv4"
	}
	
	srcIP, _ := info["src_ip"].(string)
	dstIP, _ := info["dst_ip"].(string)
	
	return fmt.Sprintf("IPv4 %s -> %s", 
		getOrDefault(srcIP, "?"),
		getOrDefault(dstIP, "?"))
}

// Helper function to return default value if string is empty
func getOrDefault(val interface{}, def string) string {
	if val == nil {
		return def
	}
	
	if s, ok := val.(string); ok && s != "" {
		return s
	}
	
	return def
}

File: packet_seaquence/log/python_shark/src/protocol_analyzer/ipv4.py
from src.protocol_analyzer.base import ProtocolAnalyzer

class AnalyzeIPv4(ProtocolAnalyzer):
    """
    IPv4内の詳細情報を解析するクラス
    """
    
    def analyze(self):
        if not hasattr(self.packet, 'ip'):
            return None
        try:
            ipv4_info = {}
            if hasattr(self.packet.ip, 'src'):
                ipv4_info['src_ip'] = self.packet.ip.src
            if hasattr(self.packet.ip, 'dst'):
                ipv4_info['dst_ip'] = self.packet.ip.dst
            if hasattr(self.packet.ip, 'ttl'):
                ipv4_info['ttl'] = self.packet.ip.ttl
            if hasattr(self.packet.ip, 'flags'):
                ipv4_info['flags'] = self.packet.ip.flags
            if hasattr(self.packet.ip, 'len'):
                ipv4_info['length'] = self.packet.ip.len
            return ipv4_info
        except Exception as e:
            print(f"IPv4解析中にエラーが発生しました: {str(e)}")
            return None
    
    def get_display_info(self):
        info = self.analyze()
        if not info:
            return "IPv4情報なし"
        return f"IPv4 {info.get('src_ip','?')} -> {info.get('dst_ip','?')} (TTL={info.get('ttl','?')})"
    
    def get_summary(self):
        info = self.analyze()
        if not info:
            return "IPv4"
        return f"IPv4 {info.get('src_ip','?')} -> {info.get('dst_ip','?')}"